[improve][monitor]Add prometheusRawMetricsProvider support

hangc0276 opened the pull request:
### Motivation
Currently, Pulsar Prometheus metric framework can generate new metrics from `PrometheusRawMetricsProvider`, and the current PrometheusMetricsProvider implementation doesn't support this interface, which leads to new created PrometheusMetricProvider instance can not integrate with the current metric system.

### Modification
Pulsar prometheusMetricsProvider supports the following features.
- Support label
- Support PrometheusRawMetricsProvider interface.

This Pr just migrated from the BookKeeper Prometheus metrics provider.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### Matching PR in the forked repository
PR in forked repository: https://github.com/hangc0276/pulsar/pull/4

hangc0276 commented:
@asafm @tjiuming  Would you please help take a look? thanks.

asafm commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java:
```diff
@@ -1634,6 +1643,29 @@ public static Multimap<String, Metric> parseMetrics(String metrics) {
+        rawMetricsProvider.getStatsLogger("test").getOpStatsLogger("test_metrics")
+            .registerSuccessfulEvent(100, TimeUnit.NANOSECONDS);
+
+        getPulsar().addPrometheusRawMetricsProvider(rawMetricsProvider);
```
If you already have integration built it, and you go about querying the real metrics endpoint, why not check that you have a real BK client metric *with* the labels you expect to have but now you actually have them?

hangc0276 replied:
The PR is not mainly developed for the BK client metrics, and it is used for the plugin metrics integrated with Pulsar broker.

asafm replied:
I don't have enough knowledge on 
> is used for the plugin metrics integrated with Pulsar broker.

Can you please elaborate and give references?

asafm replied:
Oh, I read your code below. 
I'm still amazed after spending weeks reading all the metrics code, I'm still discovering new bits.
I was wondering what was that weird `PrometheusRawMetricsProvider` nobody was actually using inside Pulsar.
So you're saying it's in fact a public API for other plugin developers to rely on that? Can you please give me a reference for such one?

hangc0276 replied:
Yes, it is a public API for other plugins which run within the Pulsar broker to expose metrics to the Pulsar broker metric port. A public plugin repo is KOP. https://github.com/streamnative/kop/blob/a9f56e9b0435429dd8d977ba948c6772e6fe5b86/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java#L280. 
I'm developing another plugin, which also needs this public API to expose metrics.

asafm replied:
@hangc0276 Ok, from my understanding, and correct me if I'm wrong here: This PR fixes the BK Metrics API implementation in such a way that it will support labels. *But* it doesn't fix another issue here: The ability to *use* `PrometheusMetricsProvider` by plugins. 

So today, and in your PR, we only print the metrics registered to `PrometheusMetricsProvider` instantiated in `ManagedLedgerFactory` - which is good as we now get the BK client metrics with labels.

*But*, I believe the correct solution should be to instantiate `PrometheusMetricsProvider` in `PulsarService`, and pass it along to `ManagedLedgerFactory` and also available to retrieve by any plugin if needed.

The way I see it, plugins should use:
1. Prometheus Client library - preferred.
2. BK Metrics API implementation - not the best-preferred way. 
3. Raw metrics provider, if all other means fail.
If we're already doing this, at least let's do this right. In your case, you know you want to use the `PrometheusMetricsProvider` maybe we should just expose it and use it properly.

asafm commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java:
```diff
@@ -1634,6 +1643,29 @@ public static Multimap<String, Metric> parseMetrics(String metrics) {
+        HttpClient httpClient = HttpClientBuilder.create().build();
+        final String metricsEndPoint = getPulsar().getWebServiceAddress() + "/metrics";
+        HttpResponse response = httpClient.execute(new HttpGet(metricsEndPoint));
+        InputStream inputStream = response.getEntity().getContent();
```
use `EntityUtils.toString`

hangc0276 replied:
done

asafm commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java:
```diff
@@ -1634,6 +1643,29 @@ public static Multimap<String, Metric> parseMetrics(String metrics) {
+        while((str = reader.readLine()) != null){
+            sb.append(str);
+        }
+        Assert.assertTrue(sb.toString().contains("test_metrics"));
```
Maybe `parseMetrics()`

hangc0276 replied:
done

hangc0276 replied:
I will use another PR to add checks for the metrics string format.

asafm commented:
> @asafm @tjiuming Would you please help take a look? thanks.

Done @hangc0276

asafm commented:
For future readers of this PR: 

**Why is this PR needed?**
Pulsar uses Bookkeeper Client, which has metrics. The way to get those metrics is to implement the BK Metric library API, which Pulsar has already done through `PrometheusMetricsProvider`, `PrometheusStatsLogger`, etc.
The problem with that implementation was that the API itself supported having labels per metric, but the implementation ignored the labels provided to it. 

This PR aims to fix and add support for the labels given to it via the API methods.
The impact is that now Pulsar `/metrics` endpoint, which exposes the Pulsar metrics in Prometheus format, will contain the Bookkeeper client metrics *with* labels (which were previously stripped).

hangc0276 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusMetricsProvider.java:
```diff
@@ -18,109 +18,109 @@
+        labels = Collections.singletonMap(CLUSTER_NAME, conf.getString(CLUSTER_NAME, DEFAULT_CLUSTER_NAME));
+
+        executor.scheduleAtFixedRate(() -> {
+            rotateLatencyCollection();
```
I add it back.

hangc0276 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusMetricsProvider.java:
```diff
@@ -18,109 +18,109 @@
     public static final String DEFAULT_CLUSTER_NAME = "pulsar";
 
-    private String cluster;
-    private final CachingStatsProvider cachingStatsProvider;
```
I add the cachingStatsProvider back with a small change.

hangc0276 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusMetricsProvider.java:
```diff
@@ -18,109 +18,109 @@
-        opStats.forEach((name, opStatLogger) -> PrometheusTextFormatUtil.writeOpStat(writer, name, cluster,
-                opStatLogger));
+        PrometheusTextFormat prometheusTextFormat = new PrometheusTextFormat();
+        PrometheusTextFormat.writeMetricsCollectedByPrometheusClient(writer, registry);
```
I removed it.

hangc0276 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusMetricsProvider.java:
```diff
@@ -18,109 +18,109 @@
-        counters.forEach((name, counter) -> PrometheusTextFormatUtil.writeCounter(writer, name, cluster, counter));
-        opStats.forEach((name, opStatLogger) -> PrometheusTextFormatUtil.writeOpStat(writer, name, cluster,
-                opStatLogger));
+        PrometheusTextFormat prometheusTextFormat = new PrometheusTextFormat();
```
Good point, updated the code.

asafm replied:
Reiterate question: The only caveat I see is this: Do you see any chance that we will have two different pairs of (scopeContext, gauge/metric/...) such that they have the same metric name, maybe the only difference is in the labels?

hangc0276 replied:
Yes, there are many cases. Such as a metric name is `publishrLatency`, and the label with a different topic name. It represents different metrics.

hangc0276 replied:
I add `CachingStatsLogger` to cache the same scopeLabel instances to avoid creating too many instances with the same scope and label name.

hangc0276 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusTextFormatUtil.java:
```diff
@@ -18,121 +18,174 @@
-                .append(success.toString()).append("\", quantile=\"")
-                .append(Double.toString(quantile)).append("\"} ")
-                .append(Double.toString(opStat.getQuantileValue(success, quantile))).append('\n');
+    public static void writeMetricsCollectedByPrometheusClient(SimpleTextOutputStream w, CollectorRegistry registry) {
```
Removed it.

hangc0276 commented:
> > @asafm @tjiuming Would you please help take a look? thanks.
> 
> Done @hangc0276

@asafm Thanks for your review, I addressed all the comments, please help take a look again, thanks a lot.

asafm commented:
> > For future readers of this PR:
> > **Why is this PR needed?** Pulsar uses Bookkeeper Client, which has metrics. The way to get those metrics is to implement the BK Metric library API, which Pulsar has already done through `PrometheusMetricsProvider`, `PrometheusStatsLogger`, etc. The problem with that implementation was that the API itself supported having labels per metric, but the implementation ignored the labels provided to it.
> > This PR aims to fix and add support for the labels given to it via the API methods. The impact is that now Pulsar `/metrics` endpoint, which exposes the Pulsar metrics in Prometheus format, will contain the Bookkeeper client metrics _with_ labels (which were previously stripped).
> 
> @asafm Thanks for your explanation. Another purpose of this Pr is to support plugin metrics integrate with Pulsar broker's metrics through the interface.
> 
> https://github.com/apache/pulsar/blob/63d4cf20e7b9c9bd24d3fcd5ba7397f0d185ce57/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java#L904-L914
> 
> https://github.com/apache/pulsar/blob/63d4cf20e7b9c9bd24d3fcd5ba7397f0d185ce57/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java#L1673-L1682

How do you provide such support in your PR - that I haven't figured out yet

asafm commented:
BTW: Force push is not good - I lose all context - I can't see only the changes you've added

asafm commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/CachingStatsProvider.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    public StatsLogger getStatsLogger(String scope) {
```
Now that we have labels, should `String` -> `ScopeContext`? since sope is just the metric prefix, and in theory you can have same metric prefix, different labels

asafm commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusMetricsProvider.java:
```diff
@@ -24,37 +24,44 @@
+ * A <i>Prometheus</i> based {@link PrometheusRawMetricsProvider} implementation.
  */
-public class PrometheusMetricsProvider implements StatsProvider {
+public class PrometheusMetricsProvider implements StatsProvider, PrometheusRawMetricsProvider {
```
I don't see a reason to implement `PrometheusRawMetricsProvider`

hangc0276 replied:
Refer to https://github.com/apache/pulsar/pull/17531#discussion_r979710232

asafm replied:
Based on my reply to the comment, I think it's redundant. We should have a single `PrometheusMetricsProvider` in `PulsarService`, and its metrics should be written explicitly in `PrometheusMetricsGenerator`

asafm commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusTextFormat.java:
```diff
@@ -0,0 +1,191 @@
+                .write(Double.toString(opStat.getSum(success))).write('\n');
+    }
+
+    public static void writeMetricsCollectedByPrometheusClient(SimpleTextOutputStream w, CollectorRegistry registry) {
```
Do you need this method?

asafm commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java:
```diff
@@ -1632,6 +1639,35 @@ public static Multimap<String, Metric> parseMetrics(String metrics) {
+            final String metricsEndPoint = getPulsar().getWebServiceAddress() + "/metrics";
+            HttpResponse response = httpClient.execute(new HttpGet(metricsEndPoint));
+            Multimap<String, Metric> metrics = parseMetrics(EntityUtils.toString(response.getEntity()));
+            if (((List<Metric>) metrics.get("test_raw_writeLatency_count"))
```
Let's extract to variable `(List<Metric>) metrics.get("test_raw_writeLatency_count"))
                .get(0)` to make it easier to read the next lines

asafm commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java:
```diff
@@ -1632,6 +1639,35 @@ public static Multimap<String, Metric> parseMetrics(String metrics) {
+                .getOpStatsLogger("writeLatency")
+                .registerSuccessfulEvent(100, TimeUnit.NANOSECONDS);
+
+            getPulsar().addPrometheusRawMetricsProvider(rawMetricsProvider);
```
Why are you doing that test?

The whole feature is about adding support of labels to BK client metrics. You changed `PrometheusMetricsProvider` to do that.
You changed the integration point of `PrometheusMetricsProvider` to `PrometheusMetricsGenerator`

I don't see any relation to `PrometheusRawMetricsProvider`.

I reiterate the last comment - you should test how suddenly BK client metrics have labels unless I'm missing something big here.

hangc0276 replied:
The BK client metrics test is `testManagedLedgerBookieClientStats` and I updated the test to adjust for this change.
https://github.com/apache/pulsar/blob/90c554d3fa0efe0a24ca9af9946efa8908ccf2dc/pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java#L982-L1038

`testRawMetricsProvider` is mainly to test the `PrometheusRawMetricsProvider`. The relation is used by `getPulsar().addPrometheusRawMetricsProvider(rawMetricsProvider);` to register RawMetricsProvider to Pulsar service.

hangc0276 commented:
> BTW: Force push is not good - I lose all context - I can't see only the changes you've added

Sorry, I just rebased the master, and it needs force push, otherwise, the push will fail

hangc0276 commented:
This PR changed the metric name for BK client, and it can only be released in the major version and needs to update the doc. /cc @Anonymitaet

asafm commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusTextFormat.java:
```diff
@@ -0,0 +1,179 @@
+        // Example:
+        // # TYPE bookie_storage_entries_count gauge
+        // bookie_storage_entries_count 519
+        writeType(w, name, "gauge");
```
Now that you have grouped the metrics together by metric name, the type needs to be written once per metric name, so called from outside.

tjiuming commented:
I just wandering that why don't we use Prometheus lib, I think this implementation is hard to maintain
